backend/inspirehep/records/api/literature.py

```diff
@@ -267,106 +269,161 @@ def add_files(self, data):
         if "deleted" in data and data["deleted"]:
             LOGGER.info("Record is deleted, skipping files.", uuid=self.id)
             return data
+        builder = LiteratureBuilder()
 
         documents = data.pop("documents", [])
         figures = data.pop("figures", [])
 
-        added_documents = self.add_documents(documents)
+        self.add_files_data(documents, builder.add_document)
+        added_documents = builder.record.get("documents")
         if added_documents:
             data["documents"] = added_documents
 
-        added_figures = self.add_figures(figures)
+        self.add_files_data(figures, builder.add_figure)
+        added_figures = builder.record.get("figures")
         if added_figures:
             data["figures"] = added_figures
-
         return data
 
-    def add_documents(self, documents):
-        builder = LiteratureBuilder()
-        for document in documents:
-            if not document.get("hidden", False):
-                new_file_data = self.add_file(**document)
-                document.update(new_file_data)
-            try:
-                builder.add_document(**document)
-            except ValueError:
-                LOGGER.exception(
-                    "Duplicated document found",
-                    recid=self.get("control_number"),
-                    uuid=self.id,
-                    document=document.get("key"),
+    def _generate_tasks(self, files, executor, add_file_func):
```

Comment:
```suggestion
    def _generate_future_tasks(self, files, executor, add_file_func):
```

Reply:
Sorry I'm not convinced to this name. It sounds weird..

Reply:
what is the ``_generate_tasks``? celery tasks? ``_geenerate_tasks`` is too generic IMHO

Reply:
Yeah, but generate_future_task is also ambiguous. Maybe _generate_files_tasks or smth?